src/main/java/com/commercetools/project/sync/product/ProductSyncer.java

```diff
@@ -150,153 +128,11 @@ private static Throwable getCompletionExceptionCause(@Nonnull final Throwable ex
-        .stream()
-        .map(AttributeContainer::getAttributes)
-        .flatMap(Collection::stream)
-        .map(Attribute::getValueAsJsonNode)
-        .map(ProductSyncer::getReferences)
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList());
-  }
-
-  @Nonnull
-  private static List<JsonNode> getReferences(@Nonnull final JsonNode attributeValue) {
-    // This will only work if the reference is not expanded, otherwise behaviour is not guaranteed.
-    return attributeValue.findParents(REFERENCE_TYPE_ID_FIELD);
-  }
-
-  @Nonnull
-  private List<JsonNode> getReferencesByTypeId(
-      @Nonnull final List<JsonNode> references, @Nonnull final String typeId) {
-    return references
-        .stream()
-        .filter(reference -> typeId.equals(reference.get(REFERENCE_TYPE_ID_FIELD).asText()))
-        .collect(Collectors.toList());
-  }
-
-  @Nonnull
-  private static Set<String> getIds(@Nonnull final List<JsonNode> references) {
-    return references.stream().map(ProductSyncer::getId).collect(toSet());
-  }
-
-  @Nonnull
-  private static String getId(@Nonnull final JsonNode ref) {
-    return ref.get(REFERENCE_ID_FIELD).asText();
-  }
-
-  @Nonnull
-  private List<Product> filterOutWithIrresolvableReferences(
-      @Nonnull final List<Product> products, @Nonnull final Map<String, String> idToKey) {
-
-    return products
-        .stream()
-        .filter(
-            product -> {
-              final Set<JsonNode> irresolvableReferences =
-                  getIrresolvableReferences(product, idToKey);
-              final boolean hasIrresolvableReferences = !irresolvableReferences.isEmpty();
-              if (hasIrresolvableReferences) {
-                LOGGER.warn(
-                    format(
-                        WITH_IRRESOLVABLE_REFS_ERROR_MSG,
-                        product.getId(),
-                        getSourceClient().getConfig().getProjectKey(),
-                        irresolvableReferences));
-              }
-              return !hasIrresolvableReferences;
-            })
-        .collect(Collectors.toList());
-  }
-
-  private Set<JsonNode> getIrresolvableReferences(
-      @Nonnull final Product product, @Nonnull final Map<String, String> idToKey) {
-
-    return getAllReferences(product)
-        .stream()
-        .filter(reference -> !idToKey.containsKey(getId(reference)))
-        .collect(toSet());
-  }
-
-  private static void replaceReferences(
-      @Nonnull final List<JsonNode> references, @Nonnull final Map<String, String> idToKey) {
-
-    references.forEach(
-        reference -> {
-          final String id = reference.get(REFERENCE_ID_FIELD).asText();
-          final String key = idToKey.get(id);
-          ((ObjectNode) reference).put(REFERENCE_ID_FIELD, key);
-        });
-  }
-
   @Nonnull
   @Override
-  protected ProductQuery getQuery() {
+  protected ProductProjectionQuery getQuery() {
     ProductSyncCustomRequest customRequest = this.productSyncCustomRequest;
```

Comment:
Do we need this additional reference of `this.productSyncCustomRequest` ?

Reply:
https://github.com/commercetools/commercetools-project-sync/pull/268/commits/c7fc0f189f96c3d0300b452da40a9f30d6e269a3